src/lib/visualization/cardinality_writer.cpp

```diff
@@ -0,0 +1,197 @@
+#include "cardinality_writer.hpp"
+
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "constant_mappings.hpp"
+#include "expression/expression_utils.hpp"
+#include "expression/lqp_subquery_expression.hpp"
+#include "expression/pqp_subquery_expression.hpp"
+#include "logical_query_plan/abstract_non_query_node.hpp"
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "operators/limit.hpp"
+#include "operators/projection.hpp"
+#include "operators/table_scan.hpp"
+
+namespace opossum {
+
+CardinalityWriter::CardinalityWriter() {
+  // We can guarantee the LQP never changes during visualization and thus avoid redundant estimations for subplans
+  _cardinality_estimator.guarantee_bottom_up_construction();
+}
+
+void CardinalityWriter::write_cardinalities(const std::vector<std::shared_ptr<AbstractLQPNode>>& lqp_roots,
+																						const std::vector<std::shared_ptr<AbstractOperator>>& pqp_roots,
+																						const std::string& benchmark_item_id){
+    _build_graph(lqp_roots);
+		_build_pqp_graph(pqp_roots);
+
+		//Assert(_lqp_nodes.size() != _pqp_nodes.size(), "LQP and PQP have differing number of nodes / differing structure. Can't match.");
+
+		// std::cout << "LQP SIZE " << _lqp_nodes.size() << std::endl;
+		// std::cout << "PQP SIZE " << _pqp_nodes.size() << std::endl;
+
+		const auto lqp_nodes_size = _lqp_nodes.size();
+		const auto pqp_nodes_size = _pqp_nodes.size();
+		
+		const auto number_nodes = std::min(_lqp_nodes.size(), _pqp_nodes.size());
+
+		for (auto node = 0u; node < number_nodes; node++) {
+
+			const auto lqp_description = _lqp_nodes[node]->description();
+			const auto pqp_description = _pqp_nodes[node]->description();
+			// std::cout << _lqp_nodes[node]->description() << std::endl;
+			// std::cout << _pqp_nodes[node]->name() << " " << _pqp_nodes[node]->description() << std::endl;
+			
+			const auto lqp_row_count = _cardinality_estimator.estimate_cardinality(_lqp_nodes[node]);
+			
+			auto pqp_row_count = NAN;
+			const auto& performance_data = *_pqp_nodes[node]->performance_data;
+
+  		if (_pqp_nodes[node]->executed() && performance_data.has_output) {
+    		pqp_row_count = performance_data.output_row_count;
+  		}
+
+			// std::cout << "LQP Rowcount: " << lqp_row_count << std::endl;
+			// std::cout << "PQP Rowcount: " << pqp_row_count << std::endl;
+
+  		std::ofstream output_file;
+  		output_file.open ("./benchmark_cardinality_estimation.csv", std::ios_base::app);
+  		output_file << benchmark_item_id << "," << lqp_description << "," << pqp_description << "," << lqp_row_count << "," << pqp_row_count << "\n";
+  		output_file.close();
```

Comment:
Write Header somewhere? maybe print it to the command line? be careful: this code is executed multiple time